cli/internal/core/engine.go

```diff
@@ -385,11 +412,131 @@ func (e *Engine) ValidatePersistentDependencies(graph *graph.CompleteGraph) erro
-	return rootPipeline.GetTask(taskID, taskName)
+// GetResolvedTaskDefinition returns a "resolved" TaskDefinition composed of one
+// turbo.json in the workspace and following any `extends` keys up. If there is
+// no turbo.json in the workspace, returns the taskDefinition from the root Pipeline.
+func (e *Engine) GetResolvedTaskDefinition(pkg *fs.PackageJSON, rootPipeline *fs.Pipeline, taskName string, taskID string) (*fs.ResolvedTaskDefinition, error) {
+	taskDefinitions, err := e.getTaskDefinitionChain(rootPipeline, pkg, taskID, taskName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Start with an empty definition
+	mergedTaskDefinition := &fs.ResolvedTaskDefinition{}
+
+	// For each of the TaskDefinitions we know of, merge them in
+	for _, bookkeepingTaskDef := range taskDefinitions {
+		taskDef := bookkeepingTaskDef.TaskDefinition
+		if bookkeepingTaskDef.HasField("Outputs") {
+			mergedTaskDefinition.Outputs = &fs.TaskOutputs{
+				Inclusions: taskDef.Outputs.Inclusions,
+				Exclusions: taskDef.Outputs.Exclusions,
+			}
+		}
+
+		mergedTaskDefinition.ShouldCache = taskDef.ShouldCache
+		if bookkeepingTaskDef.HasField("EnvVarDependencies") {
+			mergedTaskDefinition.EnvVarDependencies = taskDef.EnvVarDependencies
+		}
+
+		if bookkeepingTaskDef.HasField("TopologicalDependencies") {
+			mergedTaskDefinition.TopologicalDependencies = taskDef.TopologicalDependencies
+		}
+
+		if bookkeepingTaskDef.HasField("TaskDependencies") {
+			mergedTaskDefinition.TaskDependencies = taskDef.TaskDependencies
+		}
+
+		if bookkeepingTaskDef.HasField("Inputs") {
+			mergedTaskDefinition.Inputs = taskDef.Inputs
+		}
+
+		if bookkeepingTaskDef.HasField("OutputMode") {
+			mergedTaskDefinition.OutputMode = taskDef.OutputMode
+		}
+		mergedTaskDefinition.Persistent = taskDef.Persistent
+	}
+
+	// Set defaults if we had no config for this.
+	if mergedTaskDefinition.Outputs == nil {
+		mergedTaskDefinition.Outputs = &fs.TaskOutputs{
+			Inclusions: []string{},
+			Exclusions: []string{},
+		}
+	}
+
+	return mergedTaskDefinition, nil
+}
+
+func (e *Engine) getTaskDefinitionChain(rootPipeline *fs.Pipeline, pkg *fs.PackageJSON, taskID string, taskName string) ([]fs.BookkeepingTaskDefinition, error) {
+	// Look for the taskDefinition in the root pipeline. We'll wait to throw errors until the end
+	rootTaskDefinition, _ := rootPipeline.GetTask(taskID, taskName)
+	var workspaceDefinition *fs.BookkeepingTaskDefinition
+
+	// If the taskID is a root task (e.g. //#build), we don't need to look
+	// for a workspace task, since these can only be defined in the root turbo.json.
+	taskIDPackage, _ := util.GetPackageTaskFromId(taskID)
+	if taskIDPackage != util.RootPkgName && taskIDPackage != ROOT_NODE_NAME {
+		// Look up task definition in turbo.json in the workspace directory
+		workspaceConfigPath := turbopath.AbsoluteSystemPath(pkg.Dir).UntypedJoin("turbo.json")
```

Comment:
I think this is preferred over explicit casting
```suggestion
		workspaceConfigPath := pkg.Dir.Join("turbo.json").RestoreAnchor(e.completeGraph.RepoRoot)
```

Reply:
thanks for the suggestion, pushed from local since I was making other changes as well!